### ISSUE TITLE
Canvas pointer input

### ISSUE THREAD

**Issue by kevinross100** [Microsoft/Win2D#423]
Hi,

I am in the process of writing an application to control an audio mixing console and as such needed to create a number of custom controls that react to user input namely a fader and a dial control. When the control is not being manipulated it does not update.

I have successfully created these controls using the win2d canvas control I get the user input by capturing the Manipulation events of the canvas. So far so good every works as I want it to. My question is one of performance. When I have lots of these controls on a form (up to 32) the response gets quick slow and the control doesn’t redraw smoothly. I considered swapping to the animated canvas control but when I read the documentation it said that the control is not really suited to having multiple animated controls on screen at one time.

Is there a way I can improve the performance of the capture or should I disregard the documentation and switch to the animated control?

Thanks

**Comment by shawnhar** [comment]
CanvasControl is better than CanvasAnimatedControl for this scenario.

I recommend trying to narrow down the source of this performance drop.  Are you having latency on the input processing, or does it take too long to redraw the control in response to changes?  Are you redrawing more controls than strictly necessary?  (triggering a redraw on just one control should perform the same regardless of how many other not-redrawing controls are elsewhere on the screen at the time)

**Comment by kevinross100** [comment]
Each control handles its own drawing so changing one wouldn't cause the rest to redraw. When I check to see how long the draw takes it is about 3000 ticks (0.3ms) so I don't think it is the draw that is taking the time. Is there a faster way of capturing the input of the user pressing their finger down (manipulation starting) and dragging the fader (manipulation delta)

**Comment by shawnhar** [comment]
I don't know much about the input side of XAML, sorry.  Manipulation events certainly seem like the right way to do this sort of thing.

Timing the calls inside your draw handler isn't very meaningful as that is just queuing up work for later execution.  To see how long the actual drawing takes, add a Flush call inside the region being measured (but don't leave that in when you are done profiling, as forcing an early flush will degrade overall performance).

I'm suspicious of your observation that increasing the number of these controls hurts performance.  This makes me think you must unintentionally be doing some work in controls other than the one the user is interacting with.  Either redrawing more than you intend to, or perhaps more than one control is responding to input events at the same time?

**Comment by kevinross100** [comment]
Thank you, once I added the flush command the timing stays quite low so this further points the the manipulation events not being handled fast enough. I have checked and only one control gets redrawn when manipulated so it's not triggering anything. Would it make a difference that the controls are inside a template of an ItemsControl I wonder?

Either way I think I need to start looking at what input handling mechanism touch games would use on UWP in order to gain more performance.
